# Project: Wrangling and Analyze Data

This project makes part of the coursework leading to Udacity's Data Analysis Nanodegree (DAND). Its objective is to demonstrate skills for the Data Wrangling phase of the Data Analysis Process. Data Wrangling is a key part in analytics, and is typically described as the one where analysts and data scientists spend the most part of their time.

The project gathers data from different sources related to the WeRateDogs twitter account, which posts and rates photos of folowers' dogs. After assessing and cleaning the data, reports are written to communicate the results of an initial analysis.

In [81]:
# Import the required packages


# Import Libraries

# Data science
import pandas as pd
import tweepy as tw
import requests
import json
import os
import numpy as np

# Plotting
import matplotlib.pyplot as plt
import seaborn as sns

# Plotting Pretty figures and avoiding blurry images
%config InlineBackend.figure_format = 'retina'

# Larger scale for plots in notebooks
sns.set_context('notebook')

# Machine Learning
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.linear_model import LogisticRegression



# Ignore warnings
import warnings
warnings.filterwarnings('ignore')



## Data Gathering
In the cell below, gather **all** three pieces of data for this project and load them in the notebook. **Note:** the methods required to gather each data are different.
1. Directly download the WeRateDogs Twitter archive data (twitter_archive_enhanced.csv)
2. A file downloaded programatically, with breed predictions for each dog
3. Additional data about each tweet (retweets and fav's), retrieved from Twitter's API

### Import a File on hand into a padas dataframe

In [2]:
# Import file with twitter posts as a pandas dataframe
feed = pd.read_csv('twitter-archive-enhanced.csv')


### Dowload a file from the Internet programatically

In [3]:
# Programatically download the datafile with breed predictions based on the tweet's image from a URL to insure reproducibility
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
with open(os.path.join(url.split('/')[-1]), mode='wb') as file:
        file.write(response.content)
        
predictions = pd.read_csv('image-predictions.tsv', sep='\t')

### Retrieve additional data from Twitter APIs, for each of the tweets in twitter-archive-enhanced.csv

Each tweet's retweet count and favorite ("like") count at minimum, and any additional data you find interesting. Using the tweet IDs in the WeRateDogs Twitter archive, query the Twitter API for each tweet's JSON data using Python's Tweepy library and store each tweet's entire set of JSON data in a file called tweet_json.txt file. Each tweet's JSON data should be written to its own line. Then read this .txt file line by line into a pandas DataFrame with (at minimum) tweet ID, retweet count, and favorite count.

In [4]:
# Twitter keys (delete before sharing)
consumer_key=""
consumer_secret=""
access_token=""
access_token_secret=""

In [5]:
# query the Twitter API for each tweet's JSON data using Python's Tweepy
auth = tw.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)

api = tw.API(auth, wait_on_rate_limit = True)

In [6]:
# Dataframe with the tweet ids in twitter_archive-enhanced.csv
tweet_ids = feed.tweet_id

In [7]:
# retrieve the full JSON for the status of each id in twitter-archive-enhanced.csv
# This script takes over 35 minutes to run
# load tweets stautus json file after gathering 
tweet_status = pd.read_json('tweet_json.txt', lines = True,encoding='utf-8')
failed_id = []
with open('tweet_json.txt', 'w', encoding='UTF-8') as json_file:
    for id_ in tweet_ids:
        try:
            tweet = api.get_status(id_)
            tweet = tweet._json
            json.dump(tweet, json_file)
            json_file.write('\n')
        except:
            print('Failed retrieving tweet with id ',id_)
            failed_id.append(id_)

Failed retrieving tweet with id  892420643555336193
Failed retrieving tweet with id  892177421306343426
Failed retrieving tweet with id  891815181378084864
Failed retrieving tweet with id  891689557279858688
Failed retrieving tweet with id  891327558926688256
Failed retrieving tweet with id  891087950875897856
Failed retrieving tweet with id  890971913173991426
Failed retrieving tweet with id  890729181411237888
Failed retrieving tweet with id  890609185150312448
Failed retrieving tweet with id  890240255349198849
Failed retrieving tweet with id  890006608113172480
Failed retrieving tweet with id  889880896479866881
Failed retrieving tweet with id  889665388333682689
Failed retrieving tweet with id  889638837579907072
Failed retrieving tweet with id  889531135344209921
Failed retrieving tweet with id  889278841981685760
Failed retrieving tweet with id  888917238123831296
Failed retrieving tweet with id  888804989199671297
Failed retrieving tweet with id  888554962724278272
Failed retri

Failed retrieving tweet with id  861005113778896900
Failed retrieving tweet with id  860981674716409858
Failed retrieving tweet with id  860924035999428608
Failed retrieving tweet with id  860563773140209665
Failed retrieving tweet with id  860524505164394496
Failed retrieving tweet with id  860276583193509888
Failed retrieving tweet with id  860184849394610176
Failed retrieving tweet with id  860177593139703809
Failed retrieving tweet with id  859924526012018688
Failed retrieving tweet with id  859851578198683649
Failed retrieving tweet with id  859607811541651456
Failed retrieving tweet with id  859196978902773760
Failed retrieving tweet with id  859074603037188101
Failed retrieving tweet with id  858860390427611136
Failed retrieving tweet with id  858843525470990336
Failed retrieving tweet with id  858471635011153920
Failed retrieving tweet with id  858107933456039936
Failed retrieving tweet with id  857989990357356544
Failed retrieving tweet with id  857746408056729600
Failed retri

Failed retrieving tweet with id  834931633769889797
Failed retrieving tweet with id  834786237630337024
Failed retrieving tweet with id  834574053763584002
Failed retrieving tweet with id  834477809192075265
Failed retrieving tweet with id  834458053273591808
Failed retrieving tweet with id  834209720923721728
Failed retrieving tweet with id  834167344700198914
Failed retrieving tweet with id  834089966724603904
Failed retrieving tweet with id  834086379323871233
Failed retrieving tweet with id  833863086058651648
Failed retrieving tweet with id  833826103416520705
Failed retrieving tweet with id  833732339549220864
Failed retrieving tweet with id  833722901757046785
Failed retrieving tweet with id  833479644947025920
Failed retrieving tweet with id  833124694597443584
Failed retrieving tweet with id  832998151111966721
Failed retrieving tweet with id  832769181346996225
Failed retrieving tweet with id  832757312314028032
Failed retrieving tweet with id  832682457690300417
Failed retri

Failed retrieving tweet with id  816091915477250048
Failed retrieving tweet with id  816062466425819140
Failed retrieving tweet with id  816014286006976512
Failed retrieving tweet with id  815990720817401858
Failed retrieving tweet with id  815966073409433600
Failed retrieving tweet with id  815745968457060357
Failed retrieving tweet with id  815736392542261248
Failed retrieving tweet with id  815639385530101762
Failed retrieving tweet with id  815390420867969024
Failed retrieving tweet with id  814986499976527872
Failed retrieving tweet with id  814638523311648768
Failed retrieving tweet with id  814578408554463233
Failed retrieving tweet with id  814530161257443328
Failed retrieving tweet with id  814153002265309185
Failed retrieving tweet with id  813944609378369540
Failed retrieving tweet with id  813910438903693312
Failed retrieving tweet with id  813812741911748608
Failed retrieving tweet with id  813800681631023104
Failed retrieving tweet with id  813217897535406080
Failed retri

Failed retrieving tweet with id  793962221541933056
Failed retrieving tweet with id  793845145112371200
Failed retrieving tweet with id  793614319594401792
Failed retrieving tweet with id  793601777308463104
Failed retrieving tweet with id  793500921481273345
Failed retrieving tweet with id  793286476301799424
Failed retrieving tweet with id  793271401113350145
Failed retrieving tweet with id  793256262322548741
Failed retrieving tweet with id  793241302385262592
Failed retrieving tweet with id  793226087023144960
Failed retrieving tweet with id  793210959003287553
Failed retrieving tweet with id  793195938047070209
Failed retrieving tweet with id  793180763617361921
Failed retrieving tweet with id  793165685325201412
Failed retrieving tweet with id  793150605191548928
Failed retrieving tweet with id  793135492858580992
Failed retrieving tweet with id  793120401413079041
Failed retrieving tweet with id  792913359805018113
Failed retrieving tweet with id  792883833364439040
Failed retri

Failed retrieving tweet with id  773922284943896577
Failed retrieving tweet with id  773704687002451968
Failed retrieving tweet with id  773670353721753600
Failed retrieving tweet with id  773547596996571136
Failed retrieving tweet with id  773336787167145985
Failed retrieving tweet with id  773308824254029826
Failed retrieving tweet with id  773247561583001600
Failed retrieving tweet with id  773191612633579521
Failed retrieving tweet with id  772877495989305348
Failed retrieving tweet with id  772826264096874500
Failed retrieving tweet with id  772615324260794368
Failed retrieving tweet with id  772581559778025472
Failed retrieving tweet with id  772193107915964416
Failed retrieving tweet with id  772152991789019136
Failed retrieving tweet with id  772117678702071809
Failed retrieving tweet with id  772114945936949249
Failed retrieving tweet with id  772102971039580160
Failed retrieving tweet with id  771908950375665664
Failed retrieving tweet with id  771770456517009408
Failed retri

Failed retrieving tweet with id  752334515931054080
Failed retrieving tweet with id  752309394570878976
Failed retrieving tweet with id  752173152931807232
Failed retrieving tweet with id  751950017322246144
Failed retrieving tweet with id  751937170840121344
Failed retrieving tweet with id  751830394383790080
Failed retrieving tweet with id  751793661361422336
Failed retrieving tweet with id  751598357617971201
Failed retrieving tweet with id  751583847268179968
Failed retrieving tweet with id  751538714308972544
Failed retrieving tweet with id  751456908746354688
Failed retrieving tweet with id  751251247299190784
Failed retrieving tweet with id  751205363882532864
Failed retrieving tweet with id  751132876104687617
Failed retrieving tweet with id  750868782890057730
Failed retrieving tweet with id  750719632563142656
Failed retrieving tweet with id  750506206503038976
Failed retrieving tweet with id  750429297815552001
Failed retrieving tweet with id  750383411068534784
Failed retri

Failed retrieving tweet with id  734787690684657664
Failed retrieving tweet with id  734776360183431168
Failed retrieving tweet with id  734559631394082816
Failed retrieving tweet with id  733828123016450049
Failed retrieving tweet with id  733822306246479872
Failed retrieving tweet with id  733482008106668032
Failed retrieving tweet with id  733460102733135873
Failed retrieving tweet with id  733109485275860992
Failed retrieving tweet with id  732732193018155009
Failed retrieving tweet with id  732726085725589504
Failed retrieving tweet with id  732585889486888962
Failed retrieving tweet with id  732375214819057664
Failed retrieving tweet with id  732005617171337216
Failed retrieving tweet with id  731285275100512256
Failed retrieving tweet with id  731156023742988288
Failed retrieving tweet with id  730924654643314689
Failed retrieving tweet with id  730573383004487680
Failed retrieving tweet with id  730427201120833536
Failed retrieving tweet with id  730211855403241472
Failed retri

Failed retrieving tweet with id  709918798883774466
Failed retrieving tweet with id  709901256215666688
Failed retrieving tweet with id  709852847387627521
Failed retrieving tweet with id  709566166965075968
Failed retrieving tweet with id  709556954897764353
Failed retrieving tweet with id  709519240576036864
Failed retrieving tweet with id  709449600415961088
Failed retrieving tweet with id  709409458133323776
Failed retrieving tweet with id  709225125749587968
Failed retrieving tweet with id  709207347839836162
Failed retrieving tweet with id  709198395643068416
Failed retrieving tweet with id  709179584944730112
Failed retrieving tweet with id  709158332880297985
Failed retrieving tweet with id  709042156699303936
Failed retrieving tweet with id  708853462201716736
Failed retrieving tweet with id  708845821941387268
Failed retrieving tweet with id  708834316713893888
Failed retrieving tweet with id  708810915978854401
Failed retrieving tweet with id  708738143638450176
Failed retri

Failed retrieving tweet with id  698178924120031232
Failed retrieving tweet with id  697995514407682048
Failed retrieving tweet with id  697990423684476929
Failed retrieving tweet with id  697943111201378304
Failed retrieving tweet with id  697881462549430272
Failed retrieving tweet with id  697630435728322560
Failed retrieving tweet with id  697616773278015490
Failed retrieving tweet with id  697596423848730625
Failed retrieving tweet with id  697575480820686848
Failed retrieving tweet with id  697516214579523584
Failed retrieving tweet with id  697482927769255936
Failed retrieving tweet with id  697463031882764288
Failed retrieving tweet with id  697270446429966336
Failed retrieving tweet with id  697259378236399616
Failed retrieving tweet with id  697255105972801536
Failed retrieving tweet with id  697242256848379904
Failed retrieving tweet with id  696900204696625153
Failed retrieving tweet with id  696894894812565505
Failed retrieving tweet with id  696886256886657024
Failed retri

Failed retrieving tweet with id  687127927494963200
Failed retrieving tweet with id  687124485711986689
Failed retrieving tweet with id  687109925361856513
Failed retrieving tweet with id  687102708889812993
Failed retrieving tweet with id  687096057537363968
Failed retrieving tweet with id  686947101016735744
Failed retrieving tweet with id  686760001961103360
Failed retrieving tweet with id  686749460672679938
Failed retrieving tweet with id  686730991906516992
Failed retrieving tweet with id  686683045143953408
Failed retrieving tweet with id  686618349602762752
Failed retrieving tweet with id  686606069955735556
Failed retrieving tweet with id  686394059078897668
Failed retrieving tweet with id  686386521809772549
Failed retrieving tweet with id  686377065986265092
Failed retrieving tweet with id  686358356425093120
Failed retrieving tweet with id  686286779679375361
Failed retrieving tweet with id  686050296934563840
Failed retrieving tweet with id  686035780142297088
Failed retri

Failed retrieving tweet with id  679527802031484928
Failed retrieving tweet with id  679511351870550016
Failed retrieving tweet with id  679503373272485890
Failed retrieving tweet with id  679475951516934144
Failed retrieving tweet with id  679462823135686656
Failed retrieving tweet with id  679405845277462528
Failed retrieving tweet with id  679158373988876288
Failed retrieving tweet with id  679148763231985668
Failed retrieving tweet with id  679132435750195208
Failed retrieving tweet with id  679111216690831360
Failed retrieving tweet with id  679062614270468097
Failed retrieving tweet with id  679047485189439488
Failed retrieving tweet with id  679001094530465792
Failed retrieving tweet with id  678991772295516161
Failed retrieving tweet with id  678969228704284672
Failed retrieving tweet with id  678800283649069056
Failed retrieving tweet with id  678798276842360832
Failed retrieving tweet with id  678774928607469569
Failed retrieving tweet with id  678767140346941444
Failed retri

Failed retrieving tweet with id  674739953134403584
Failed retrieving tweet with id  674737130913071104
Failed retrieving tweet with id  674690135443775488
Failed retrieving tweet with id  674670581682434048
Failed retrieving tweet with id  674664755118911488
Failed retrieving tweet with id  674646392044941312
Failed retrieving tweet with id  674644256330530816
Failed retrieving tweet with id  674638615994089473
Failed retrieving tweet with id  674632714662858753
Failed retrieving tweet with id  674606911342424069
Failed retrieving tweet with id  674468880899788800
Failed retrieving tweet with id  674447403907457024
Failed retrieving tweet with id  674436901579923456
Failed retrieving tweet with id  674422304705744896
Failed retrieving tweet with id  674416750885273600
Failed retrieving tweet with id  674410619106390016
Failed retrieving tweet with id  674394782723014656
Failed retrieving tweet with id  674372068062928900
Failed retrieving tweet with id  674330906434379776
Failed retri

Failed retrieving tweet with id  671390180817915904
Failed retrieving tweet with id  671362598324076544
Failed retrieving tweet with id  671357843010908160
Failed retrieving tweet with id  671355857343524864
Failed retrieving tweet with id  671347597085433856
Failed retrieving tweet with id  671186162933985280
Failed retrieving tweet with id  671182547775299584
Failed retrieving tweet with id  671166507850801152
Failed retrieving tweet with id  671163268581498880
Failed retrieving tweet with id  671159727754231808
Failed retrieving tweet with id  671154572044468225
Failed retrieving tweet with id  671151324042559489
Failed retrieving tweet with id  671147085991960577
Failed retrieving tweet with id  671141549288370177
Failed retrieving tweet with id  671138694582165504
Failed retrieving tweet with id  671134062904504320
Failed retrieving tweet with id  671122204919246848
Failed retrieving tweet with id  671115716440031232
Failed retrieving tweet with id  671109016219725825
Failed retri

Failed retrieving tweet with id  668587383441514497
Failed retrieving tweet with id  668567822092664832
Failed retrieving tweet with id  668544745690562560
Failed retrieving tweet with id  668542336805281792
Failed retrieving tweet with id  668537837512433665
Failed retrieving tweet with id  668528771708952576
Failed retrieving tweet with id  668507509523615744
Failed retrieving tweet with id  668496999348633600
Failed retrieving tweet with id  668484198282485761
Failed retrieving tweet with id  668480044826800133
Failed retrieving tweet with id  668466899341221888
Failed retrieving tweet with id  668297328638447616
Failed retrieving tweet with id  668291999406125056
Failed retrieving tweet with id  668286279830867968
Failed retrieving tweet with id  668274247790391296
Failed retrieving tweet with id  668268907921326080
Failed retrieving tweet with id  668256321989451776
Failed retrieving tweet with id  668248472370458624
Failed retrieving tweet with id  668237644992782336
Failed retri

In [88]:
# Read the contents of the created file and create a dataframe with the fields of interest 
# Reading gathered files
###more_data = pd.read_json('tweet_json.txt', lines = True,encoding='utf-8')

df = []
with open('tweet_json.txt') as f:
    for line in f:
        tweet = (json.loads(line))
        tweet_id = tweet['id']
        retweet_count = tweet['retweet_count']
        favorite_count = tweet['favorite_count']
        create_date = tweet['created_at']
        df.append({'retweet_count' : retweet_count,
                  'favorite_count' : favorite_count,
                  'create_date' : create_date,
                  'tweet_id' : tweet_id})
        
more_data = pd.DataFrame(df, columns = ['tweet_id', 'retweet_count', 'favorite_count', 'create_date'])

# Assess

Assessing is the process of checking the data sets for two things: quality issues and lack of tidiness. Data that has quality issues has problems with its contents: missing, corrupted, inacurate, duplicate or incorrect data. This is called "dirty data". In the contrary, "untidy" or "messy" data has specific structural issues.

You look for these issues in two ways: visually and programatically. When you detect an issue you document it, to make cleaning easier. You don't write yet how to clean the data, but only state the observation.

Visual assessment is great for getting acquainted with the data and understanding what it is all about. That's what we will do now:

In [9]:
# assessing the file with the retrieved data for WeRateDogs tweets
feed.head()


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [10]:
# assessing the file with the breed predictions for each dog
predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [89]:
more_data.head()

tweet_id  retweet_count  favorite_count  \
0  892420643555336193           8323           38047   
1  892177421306343426           6149           32665   
2  891815181378084864           4069           24579   
3  891689557279858688           8469           41402   
4  891327558926688256           9164           39579   

                      create_date  
0  Tue Aug 01 16:23:56 +0000 2017  
1  Tue Aug 01 00:17:27 +0000 2017  
2  Mon Jul 31 00:18:03 +0000 2017  
3  Sun Jul 30 15:58:51 +0000 2017  
4  Sat Jul 29 16:00:24 +0000 2017

### Programatical Assessment
After assessing visually the three datasets and identifying several tidiness issues, we now go deeper using Programatic Assessment:

In [90]:
feed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

The missing data for retweets and replies is fine, since we are not interested in those. 59 expanded URLs are missing.

In [91]:
predictions.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   object 
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(1), object(5)
memory usage: 152.1+ KB


In [92]:
more_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2340 entries, 0 to 2339
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   tweet_id        2340 non-null   int64 
 1   retweet_count   2340 non-null   int64 
 2   favorite_count  2340 non-null   int64 
 3   create_date     2340 non-null   object
dtypes: int64(3), object(1)
memory usage: 73.2+ KB


In [93]:
# We test whether the missing expanded URLs are a problem.
# If there is a link in the tweet's text and no expanded URL then there is an issue

explore = feed[feed['expanded_urls'].isnull()]
explore['text'].str.contains('http').sum()

0

No issue there. Every tweet with an url has an expanded URL. Those without one are usually retweets or replies.

In [94]:
# value counts for the source column
feed.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

In [95]:
# every numerator not other than 10-14 in the dogs rating is a possible quality issue
feed.rating_numerator.value_counts().sort_index(ascending=False)

1776      1
960       1
666       1
420       2
204       1
182       1
165       1
144       1
143       1
121       1
99        1
88        1
84        1
80        1
75        2
60        1
50        1
45        1
44        1
27        1
26        1
24        1
20        1
17        1
15        2
14       54
13      351
12      558
11      464
10      461
9       158
8       102
7        55
6        32
5        37
4        17
3        19
2         9
1         9
0         2
Name: rating_numerator, dtype: int64

In [96]:
# every denominator not equal to 10 in the dogs rating is a possible quality issue
feed.rating_denominator.value_counts().sort_index(ascending=False)

170       1
150       1
130       1
120       1
110       1
90        1
80        2
70        1
50        3
40        1
20        2
16        1
15        1
11        3
10     2333
7         1
2         1
0         1
Name: rating_denominator, dtype: int64

In [97]:
# we check for duplicated IDs in the feeds dataframe
feed[feed.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

There are no duplicated tweets in the dataset, testing by its ID

In [98]:
predictions.describe()

img_num      p1_conf       p2_conf       p3_conf
count  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean      1.203855     0.594548  1.345886e-01  6.032417e-02
std       0.561875     0.271174  1.006657e-01  5.090593e-02
min       1.000000     0.044333  1.011300e-08  1.740170e-10
25%       1.000000     0.364412  5.388625e-02  1.622240e-02
50%       1.000000     0.588230  1.181810e-01  4.944380e-02
75%       1.000000     0.843855  1.955655e-01  9.180755e-02
max       4.000000     1.000000  4.880140e-01  2.734190e-01

The predictions dataframe is the only one with numerical data. With .describe() we can check whether the ranges, means and quartiles make sense. In this test, we correctly see that the probability for the first guess has the largest value. We also see that the highest probability is equal to 1, which is expected.

In [99]:
# Checking for duplicated columns by creating a list of every column in every dataset
all_columns = pd.Series(list(feed) + list(predictions) + list(more_data))
# all_columns
all_columns[all_columns.duplicated()]

17    tweet_id
29    tweet_id
dtype: object


The only duplicated column is tweet_id, which is ok. This is how tidiness is assessed programatically.

## Issues


Quality

`twitter-archive-enhanced.csv`

- The extracted rating numerators are wrong when they are decimals like 13.5. Status:solved.
- source column has innecesary HTML code. Status: solved
- some tweets are retweets or replies. Status: solved
- timestamp column is not in datatime format. Status: solved
- The rating for dog in tweet with id 716439118184652801, 722974582966214656 and 682962037429899265 are wrong. Status: solved
- Many dog names are wrong, including hose with id 740373189193256964 and 770414278348247044. Many names are just other words in English. Status: solved
- The dog stage for dog in tweet with id 854010172552949760 is wrong. Status: solved
- some rows have several identical values in the expanded_url column concatenated by a comma. Status: solved
- Tweet_id fields in the three datasets are stored as numeric values and should be strings. Status: solved
- Predictions are spread in three columns. Status: solved
- Confidence intervals are spread in three columns. Status: solved
- Dog tests are spread in three columns. Status: solved
`additional twitter data`

`more_data` dataframe

- Retweet and favorite count columns have float as datatype and should be int. Status: solved


### Tidiness

`twitter-archive-enhanced.csv`

- Dog "stage" is spread in four columns. Status: solved
- Data about dogs and tweets in a same dataset. Status: solved
- The dog's raiting appears both in the "text" column and the rating numerator/denominator columns. Status: will not be solved here
- The link to the dog's photo is part of the status text and should have own column. Status: will not be solved here
`image-predictions.tsv`



- Create date exists already in twitter-archive-enhanced.tsv. Status: solved
- The dataset is split from the main dataset describing tweets. Status: solved

## Clean
Cleaning the data is the third and last step in the data wrangling process. The issues that were identified are solved programatically, in three steps: Define, Code and Test. We begin by creating copies of each of our three datasets

In [100]:
# create copies of the datasets
feed_clean = feed.copy()
predictions_clean = predictions.copy()
more_data_clean = more_data.copy()

### Some tweets are retweets or replies

## Define

Remove all observations in the feed_clean dataset that have values in the rows in_reply_to_status_id or retweeted_status_id. Then remove those two columns plus retweeted_status_user_id, retweeted_status_timestamp and in_reply_to_user_id

## Code

In [101]:
feed_clean = feed_clean[(feed_clean['in_reply_to_status_id'].isna() == True)  & (feed_clean['retweeted_status_id'].isna() == True)]

feed_clean = feed_clean.drop(['in_reply_to_status_id',
                              'in_reply_to_user_id',
                              'retweeted_status_id',
                              'retweeted_status_user_id',
                              'retweeted_status_timestamp'],
                            axis = 1)

## Test

In [102]:
feed_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   tweet_id            2097 non-null   int64 
 1   timestamp           2097 non-null   object
 2   source              2097 non-null   object
 3   text                2097 non-null   object
 4   expanded_urls       2094 non-null   object
 5   rating_numerator    2097 non-null   int64 
 6   rating_denominator  2097 non-null   int64 
 7   name                2097 non-null   object
 8   doggo               2097 non-null   object
 9   floofer             2097 non-null   object
 10  pupper              2097 non-null   object
 11  puppo               2097 non-null   object
dtypes: int64(3), object(9)
memory usage: 213.0+ KB


The extracted rating numerators are wrong when they are decimals like 13.5.

## Define


Change datatype and use regex to extract the ratings numerators again.

## Code

In [103]:
feed_clean['rating_numerator'] = feed_clean['rating_numerator'].astype('float')

In [104]:
feed_clean['rating_numerator'] = feed_clean['text'].str.extract('(\d+\.*\d*\/\d+)', expand=False).str.split('/').str[0]

## Test

In [105]:
print(feed_clean.rating_numerator.dtype)

object


In [106]:
feed_clean.loc[feed_clean['tweet_id'] == 883482846933004288, 'rating_numerator']

45    13.5
Name: rating_numerator, dtype: object

### Data about dogs and tweets in a same dataset

## Define
In order to have observational units in their own table, we need to split the feed_clean data-set into a tweets table and a dogs table. This is a design decision because there are many several tweets that include information and ratings for more than one dog. Ratings and dog stages refer to dogs, whereas retweets an likes are related to the tweets.

## Code

In [107]:
dogs = feed_clean[['tweet_id', 'name', 'doggo', 'floofer', 'pupper', 'puppo', 'rating_numerator', 'rating_denominator']].copy()
tweets = feed_clean.drop(['name', 'doggo', 'floofer', 'pupper', 'puppo', 'rating_numerator', 'rating_denominator'], axis=1)

## Test

In [108]:
print(tweets.columns)
print(dogs.columns)

Index(['tweet_id', 'timestamp', 'source', 'text', 'expanded_urls'], dtype='object')
Index(['tweet_id', 'name', 'doggo', 'floofer', 'pupper', 'puppo',
       'rating_numerator', 'rating_denominator'],
      dtype='object')


### Dog "stage" is spread in four columns

## Define
We have values as variables in the now called dogs table. We will melt the columns doggo, floofer, pupper, puppo into a new variable called dog_stage. Before melting, create a new column to identify those dogs with dog stage 'unknown'

## Code

In [109]:
# we want to create first an extra column to tag those dogs without a recognized stage. 
# This will help cleaning the dataset afterwards
def u(row):
    if row ['doggo'] == 'None' and row ['floofer'] == 'None' and row ['pupper'] == 'None' and row ['puppo'] == 'None':
        val = 'unknown'
    else:
        val ='None'
    return val

dogs['unknown'] = dogs.apply(u, axis=1)

In [110]:
dogs.sample()

tweet_id  name doggo floofer pupper puppo rating_numerator  \
523  809448704142938112  None  None    None   None  None               12   

     rating_denominator  unknown  
523                  10  unknown

In [111]:
# we melt the different dog stages into a new column calle 'dog_stage'
dogs = pd.melt(dogs, id_vars =['tweet_id', 'name', 'rating_numerator','rating_denominator'],
                     value_vars = ['doggo', 'floofer', 'pupper', 'puppo', 'unknown'],
                     var_name = 'dog_stage', 
                    value_name = 'value')

In [112]:
# clean the duplicated rows created in the preious process and drop the 'value' variable
dogs = dogs[dogs['value']!= 'None']
dogs = dogs.drop('value', axis=1)

## Test

In [113]:
dogs.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2108 entries, 9 to 10484
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   tweet_id            2108 non-null   int64 
 1   name                2108 non-null   object
 2   rating_numerator    2108 non-null   object
 3   rating_denominator  2108 non-null   int64 
 4   dog_stage           2108 non-null   object
dtypes: int64(2), object(3)
memory usage: 98.8+ KB


It is normal that now we have more records (2,108 vs 2,097), because some tweets referred to more than one dog and included more than one dog stage

### Predictions are spread in three columns

## Define
The different predictions, the confidence and the test whether the prediction is a dog or not need to be melted into four columns: prediction_number, prediction, confidence and dog

## Code

In [114]:
# This script iterates the melting process for the three categories of data: prediction, confidence and dog test. 
# at the end we have a clean tidy dataset.

# create 'prediction_number' and 'prediction' columns
predictions_clean = pd.melt(predictions_clean, id_vars = ['tweet_id', 'jpg_url', 'img_num', 'p1_conf', 'p1_dog',
       'p2_conf', 'p2_dog', 'p3_conf', 'p3_dog'],
               var_name = 'prediction_number',
               value_name = 'prediction')

# create 'confidence' column
predictions_clean = pd.melt(predictions_clean, id_vars = ['tweet_id', 'jpg_url', 'img_num', 'p1_dog',
       'p2_dog', 'p3_dog', 'prediction_number', 'prediction'],
               var_name = 'to_delete',
               value_name = 'confidence')

# remove newly created duplicated rows
predictions_clean = predictions_clean[predictions_clean['prediction_number'] == predictions_clean['to_delete'].str[:2]]
# remove unnecesary column
predictions_clean = predictions_clean.drop('to_delete', axis=1)

# create 'dog' column
predictions_clean = pd.melt(predictions_clean, id_vars = ['tweet_id', 'jpg_url', 'img_num','prediction_number', 'prediction', 'confidence'],
               var_name = 'to_delete2',
               value_name = 'dog')

# remove newly created duplicated rows
predictions_clean = predictions_clean[predictions_clean['prediction_number'] == predictions_clean['to_delete2'].str[:2]]

# remove unnecesary column
predictions_clean = predictions_clean.drop('to_delete2', axis=1)
# remove 'p' from prediction number
predictions_clean['prediction_number'] = predictions_clean['prediction_number'].str[1]


## Test

In [115]:
predictions_clean.sort_values(by = ['tweet_id', 'prediction_number']).head(9)

tweet_id                                          jpg_url  \
0      666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
8300   666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
16600  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1      666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
8301   666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
16601  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2      666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
8302   666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
16602  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

       img_num prediction_number              prediction  confidence   dog  
0            1                 1  Welsh_springer_spaniel    0.465074  True  
8300         1                 2                  collie    0.156665  True  
16600        1                 3       Shetland_sheepdog    0.061428  True  
1            1                 1                 redbone    0.506826  True  
8301         1                 2      miniature_pinscher    0.074192  True  
16601        1                 3     Rhodesian_ridgeback    0.072010  True  
2            1                 1         German_shepherd    0.596461  True  
8302         1                 2                malinois    0.138584  True  
16602        1                 3              bloodhound    0.116197  True

### Create date not needed in more_data_clean (field already exists in tweets dataframe)

## Define
Remove create_date from more_data_clean dataframe

## Code

In [116]:
more_data_clean = more_data_clean.drop('create_date', axis = 1)

## Test

In [117]:
more_data_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2340 entries, 0 to 2339
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2340 non-null   int64
 1   retweet_count   2340 non-null   int64
 2   favorite_count  2340 non-null   int64
dtypes: int64(3)
memory usage: 55.0 KB


### The more_data dataset is split from the main dataset describing tweets

## Define
Merge `more_data_clean` into `tweets`

## Code

In [118]:
# merge both dataframes on 'tweet_id'. Left join to preserve all tweets, independent of whether retweet info
# could be retrieved through twitter's API
tweets = tweets.merge(more_data_clean, how = 'left', on = 'tweet_id')

## Test


In [119]:

tweets.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2096
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   tweet_id        2097 non-null   int64  
 1   timestamp       2097 non-null   object 
 2   source          2097 non-null   object 
 3   text            2097 non-null   object 
 4   expanded_urls   2094 non-null   object 
 5   retweet_count   2095 non-null   float64
 6   favorite_count  2095 non-null   float64
dtypes: float64(2), int64(1), object(4)
memory usage: 131.1+ KB


### source column has innecesary HTML code

## Define
Remove the HTML from the source column in the feed dataframe, leaving only the clear name of the source. Use Beautiful Soup because the text is the contents of an HTML tag.

## Code

In [120]:
# Iterate through each row and extract the source's text with beautiful soup

from bs4 import BeautifulSoup as bs

new_source = []
for line, row in tweets.iterrows():
    soup = bs(row.source)
    x = soup.find('a').contents[0]
    new_source.append(x)
    
tweets['source'] = new_source

## Test

In [121]:
pd.Series(new_source).value_counts()

Twitter for iPhone     1964
Vine - Make a Scene      91
Twitter Web Client       31
TweetDeck                11
dtype: int64

### timestamp column is not in datatime format

## Define
Change timestamp datatype in tweets dataframe from object to datatime type

## Code

In [122]:
tweets['timestamp'] = pd.to_datetime(tweets['timestamp'])

## Test

In [123]:
tweets.dtypes

tweet_id                        int64
timestamp         datetime64[ns, UTC]
source                         object
text                           object
expanded_urls                  object
retweet_count                 float64
favorite_count                float64
dtype: object

- The rating for dog in tweet with id 716439118184652801 is wrong
- The rating for dog in tweet with id 722974582966214656 is wrong
- The rating for dogs in tweet with id 682962037429899265 is wrong

## Define
Look up for rating in tweet's text and update rating in dogs table

## Code

In [124]:
dogs.loc[dogs['tweet_id'] == 716439118184652801, 'rating_numerator'] = 11
dogs.loc[dogs['tweet_id'] == 716439118184652801, 'rating_denominator'] = 10

dogs.loc[dogs['tweet_id'] == 722974582966214656, 'rating_numerator'] = 13
dogs.loc[dogs['tweet_id'] == 722974582966214656, 'rating_denominator'] = 10

dogs.loc[dogs['tweet_id'] == 682962037429899265, 'rating_numerator'] = 10
dogs.loc[dogs['tweet_id'] == 682962037429899265, 'rating_denominator'] = 10

## Test

In [125]:
ids = [716439118184652801, 722974582966214656, 682962037429899265]
dogs.loc[dogs['tweet_id'].isin(ids), ['tweet_id','rating_numerator', 'rating_denominator']]

tweet_id rating_numerator  rating_denominator
9336  722974582966214656               13                  10
9373  716439118184652801               11                  10
9814  682962037429899265               10                  10

### Many dog names are wrong, including hose with id 740373189193256964 and 770414278348247044. Many names are just other words in English

## Define
Look up and correct the dog's names in the dogs table

## Code

In [126]:
dogs.loc[dogs['tweet_id'] == 887517139158093824, 'name'] = 'None'
dogs.loc[dogs['tweet_id'] == 770414278348247044, 'name'] = 'Al Cabone'

## Test

In [127]:
ids = [887517139158093824, 770414278348247044]
dogs.loc[dogs['tweet_id'].isin(ids), ['tweet_id','name']]

tweet_id       name
8409  887517139158093824       None
9020  770414278348247044  Al Cabone

### The dog stage for dog in tweet with id 854010172552949760 is wrong

## Define
Look up stage and correct stage in dogs table

## Code

In [128]:
dogs.loc[dogs['tweet_id'] == 854010172552949760, 'dog_stage'] = 'floofer'
dogs.drop(2258, inplace = True)

## Test

In [129]:
dogs.loc[dogs['tweet_id'] == 854010172552949760 , ['tweet_id','dog_stage']]

tweet_id dog_stage
161  854010172552949760   floofer

## Define

Split value to keep only the first URL in the column expanded_urls in the tweets dataset. Rename column to expanded_url

## Code

In [130]:
tweets['expanded_urls'] = tweets['expanded_urls'].str.split(',', expand=True)[0]

tweets = tweets.rename(index=str, columns={"expanded_urls": "expanded_url"})

## Test

In [131]:
tweets.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2097 entries, 0 to 2096
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype              
---  ------          --------------  -----              
 0   tweet_id        2097 non-null   int64              
 1   timestamp       2097 non-null   datetime64[ns, UTC]
 2   source          2097 non-null   object             
 3   text            2097 non-null   object             
 4   expanded_url    2094 non-null   object             
 5   retweet_count   2095 non-null   float64            
 6   favorite_count  2095 non-null   float64            
dtypes: datetime64[ns, UTC](1), float64(2), int64(1), object(3)
memory usage: 131.1+ KB


In [132]:
tweets['expanded_url'].str.contains(',').sum()

0

## Define

Change the numeric datatypes for the tweet_id variable from int to object in the three datasets

## Code

In [133]:
tweets['tweet_id'] = tweets['tweet_id'].astype(str)
dogs['tweet_id'] = dogs['tweet_id'].astype(str)
predictions['tweet_id'] = predictions['tweet_id'].astype(str)

## Test

In [134]:
print(tweets.tweet_id.dtype)
print(dogs.tweet_id.dtype)
print(predictions.tweet_id.dtype)

object
object
object


### Retweet and favorite count are in float format.

## Define
Change the datatypes for the retweet_count and favorite_count variables from float to int in tweets dataset

## Code

In [135]:
# I need to start by changing NaNs into 0, for the dtype conversion to int to work
tweets['retweet_count'] = tweets['retweet_count'].fillna(0).astype(int)
tweets['favorite_count'] = tweets['favorite_count'].fillna(0).astype(int)
tweets['retweet_count'] = tweets['retweet_count'].astype(int)
tweets['favorite_count'] = tweets['favorite_count'].astype(int)

## Test

In [136]:
tweets.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2097 entries, 0 to 2096
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype              
---  ------          --------------  -----              
 0   tweet_id        2097 non-null   object             
 1   timestamp       2097 non-null   datetime64[ns, UTC]
 2   source          2097 non-null   object             
 3   text            2097 non-null   object             
 4   expanded_url    2094 non-null   object             
 5   retweet_count   2097 non-null   int64              
 6   favorite_count  2097 non-null   int64              
dtypes: datetime64[ns, UTC](1), int64(2), object(4)
memory usage: 131.1+ KB


### The values for the column "floffer" should not be capitalized to maintain integrity with the others

## Define
Apply string methods to un-capitalize Floffer in the dog_stage values of the dogs dataset

## Code

In [137]:
dogs['dog_stage'] = dogs['dog_stage'].str.lower()

## Test

In [138]:
dogs['dog_stage'].value_counts()

unknown    1761
pupper      230
doggo        82
puppo        24
floofer      10
Name: dog_stage, dtype: int64

This concludes my cleaning of the identified issues for the project's datasets. As per the project's guidelines, the end result does not to solve every issue in the data (there are many more), but only 8 qulity issues and 2 tidiness issues. The goal is to demonstrate data wrangling skills and not to achieve a perfectly clean master dataset. After the completion with the cleaning phase, all that is left to be done is to save the cleaned master datasets, which in this case will be three tables with independent observational units (dogs.csv, tweets.csv and predictions.csv), and a master dataset (twitter_archive_master.csv) that joins the tweets data to the dogs data. Please note that the separator will be set to ';', which is the default for the locale in Germany.

In [139]:
# save the three tables with independent observational units
tweets.to_csv('tweets.csv', sep=';', index=False)
dogs.to_csv('dogs.csv', sep=';', index=False)
predictions.to_csv('predictions.csv', sep=';', index=False)

In [140]:
# create and save a master dataset. The key needs to be set to 'tweet_id' in order for the join to work.
twitter_archive_master = dogs.set_index('tweet_id').join(tweets.set_index('tweet_id'), on='tweet_id', how='left')
twitter_archive_master.to_csv('twitter_archive_master.csv', sep=';', index=False)

## Analysis
With a cleaned dataset we can now proceed to perform some basic analysis to obtain insights from the data. I will explore the data to try and solve the following questions:

- Are retweets and replies correlated?
- Which are the most common dog stage?
- What dog stage receives more likes and retweets?
- What dog stage receives the best ratings?
- What's the distribution for the retweet count?

In [141]:
import altair as alt

In [142]:
tweets_master = pd.read_csv('twitter_archive_master.csv', sep=";")

In [143]:
chart = alt.Chart(tweets).mark_point()

In [144]:
alt.Chart(tweets_master[tweets_master['dog_stage']!= 'unknown']).mark_point(opacity = 0.5).encode(
    alt.X('retweet_count',scale=alt.Scale(type='log'), bin=True),
    alt.Y('favorite_count'), 
    color = 'dog_stage').properties(title = 'Retweets and Likes are strongly correlated')

alt.Chart(...)

In [145]:
alt.Chart(tweets_master[tweets_master['dog_stage']!= 'unknown']).mark_bar(color='green', opacity=0.5).encode(
    alt.Y('dog_stage', axis=alt.Axis(title='Dog Stage')),
    x='count()'
).properties(title = 'Most dogs rated are puppers')

alt.Chart(...)

In [146]:
alt.Chart(tweets_master[tweets_master['dog_stage']!= 'unknown']).mark_bar(color='firebrick', opacity=0.5).encode(
    alt.Y('dog_stage', axis=alt.Axis(title='Dog Stage')),
    alt.X('average(favorite_count)', axis=alt.Axis(title='Average Like Count'))
).properties(title = 'People really like puppos')

alt.Chart(...)

In [147]:
alt.Chart(tweets_master[tweets_master['dog_stage']!= 'unknown']).mark_bar(color='firebrick', opacity=0.5).encode(
    alt.Y('dog_stage', axis=alt.Axis(title='Dog Stage')),
    alt.X('average(retweet_count)', axis=alt.Axis(title='Average Retweet Count'))
).properties(title = 'People really like puppos')

alt.Chart(...)

In [148]:
alt.Chart(tweets_master[tweets_master['dog_stage']!= 'unknown']).mark_bar(color='firebrick', opacity=0.5).encode(
    alt.Y('dog_stage', axis=alt.Axis(title='Dog Stage')),
    alt.X('average(rating_numerator)', axis=alt.Axis(title='Average Rating'))
).properties(title = 'Best rated dogs are puppos')

alt.Chart(...)

In [149]:
base = alt.Chart(tweets_master)

bar = base.mark_bar(color='lightblue', opacity=0.9).encode(
    alt.X('favorite_count', axis=alt.Axis(title='Likes'),
          bin=alt.Bin(maxbins=100, extent=([0,50000]))),
    y='count()'
).properties(title = 'Followers engage a lot with almost every WeRateDogs tweet ')

rule = base.mark_rule(color='red').encode(
    x='mean(favorite_count):Q',
    size=alt.value(2)
)

bar + rule

alt.LayerChart(...)

In [150]:
base = alt.Chart(tweets_master)

bar = base.mark_bar(color='lightblue', opacity=0.9).encode(
    alt.X('retweet_count', axis=alt.Axis(title='Retweets'),
          bin=alt.Bin(maxbins=100, extent=([0,16000]))),
    y='count()'
).properties(title = 'Followers engage a lot with almost every WeRateDogs tweet ')

rule = base.mark_rule(color='red').encode(
    x='mean(retweet_count):Q',
    size=alt.value(2)
)

bar + rule

alt.LayerChart(...)

## Reflexion

I thought Data Wrangling was going to be the easiest and fastest of all lessons in the DAND. However, when I started the project, I quickly realized that I needed to put much more work on it. Already in the gathering phase, specifically querying Twitter’s API, it was clear that I needed to go over the material again, take notes, and practice in my own Jupyter Notebook. I found useful the structure (define, code, test) proposed for the cleaning process, but in my clase it was clear that I needed flexibility and iteration. I had to go back to cleaning more, even until the last chart was produced. I believe sticking to the process is a must, especially when dealing with multiple sources of messy and untidy data. Although it was recommended to start by solving tidiness issues, in this case removing first the unnecessary rows with retweets and replies, followed by the organisation in tables of observational units was important. When cleaning the datasets, I felt like I was being able to put into practice many of the skills I have been learning over the last months. Extracting HTML contents from a tag within the column of a pandas dataframe using BeautifulSoup was a nice accomplishment. I have been working with relational databases for a while now, and I had already developed the sense to understand when observational units are not in their own tables. I ended up keeping separate tables for tweets, dogs and predictions. What I wasn’t sure about was whether a master dataset made sense because having different row counts makes the result different as what you expect. The key is to keep in mind whether you want to keep dogs or tweets as main organisational unit in a master dataset. When creating charts for the last section of the project, I really missed R’s ggplot. I started using matplotlib but was quickly drawn to look for other options, which brought me to the Altair library. I read its documentation and plotted my visualizations on it. Although I found Altair intuitive to use, when I looked under the hood I realized that every chart makes a copy of the whole dataset in JSON format, which makes it less than ideal when working with large datasets or when performing EDA with tons of visualizations.